### ISSUE TITLE
CI is red

### ISSUE THREAD

**Issue by peterhuene** [bytecodealliance/wasmtime-dotnet#173]
There were [breaking changes upstream](https://github.com/bytecodealliance/wasmtime/pull/5149) that impacted the C API, which are breaking the .NET CI tests.

**Comment by kpreisser** [comment]
Hi, I'm not sure if I fully understand these changes (I don't know Rust), but I gather that host traps that are created with `wasmtime_trap_new` (and e.g. a WASI exit status set by WASI's `proc_exit`) are now returned as error instead of as trap, so e.g. `wasmtime_func_call` returns an error code, which causes a `WasmtimeException` to be thrown instead of a `TrapException`. Only traps that are caused in wasm code (that map to a `wasmtime_trap_code_enum` value) are returned as actual traps.

However, I noticed that for test `Wasmtime.Tests.FuelConsumptionTests.ItThrowsOnCallingImportMethodIfNoFuelAdded`, it seems that a trap (`wasm_trap_t*`) is returned, but calling `wasmtime_trap_code` fails:
```
 Wasmtime.Tests.FuelConsumptionTests.ItThrowsOnCallingImportMethodIfNoFuelAdded
   Source: FuelConsumptionTests.cs line 126
   Duration: 160 ms

  Message: 
Expected a <Wasmtime.TrapException> to be thrown, but found <System.Runtime.InteropServices.SEHException>: "
"System.Runtime.InteropServices.SEHException with message "External component has thrown an exception."
     at Wasmtime.TrapException.Native.wasmtime_trap_code(IntPtr trap, TrapCode& exitCode)
     at Wasmtime.TrapAccessor.get_TrapCode() in C:\Users\developer4\Desktop\WebAssembly-Tests\wasmtime-dotnet\src\TrapException.cs:line 128
     at Wasmtime.TrapException.FromOwnedTrap(IntPtr trap, Boolean delete) in C:\Users\developer4\Desktop\WebAssembly-Tests\wasmtime-dotnet\src\TrapException.cs:line 306
     at Wasmtime.Function.Invoke(ReadOnlySpan`1 arguments) in C:\Users\developer4\Desktop\WebAssembly-Tests\wasmtime-dotnet\src\Function.cs:line 2141
     at Wasmtime.Function.Invoke() in C:\Users\developer4\Desktop\WebAssembly-Tests\wasmtime-dotnet\src\Function.cs:line 2082
     at Wasmtime.Tests.FuelConsumptionTests.<>c__DisplayClass19_0.<ItThrowsOnCallingImportMethodIfNoFuelAdded>b__0() in C:\Users\developer4\Desktop\WebAssembly-Tests\wasmtime-dotnet\tests\FuelConsumptionTests.cs:line 131
     at FluentAssertions.Specialized.ActionAssertions.InvokeSubject()
     at FluentAssertions.Specialized.DelegateAssertions`2.InvokeSubjectWithInterception()
.
```

Regarding the host errors, I think this means that the functionality from #172 would need to be shifted from `TrapException` to `WasmtimeException`, so that the cause can then be stored as inner exception of the `WasmtimeException`.
(Another possibility would be to directly rethrow the stored exception in such a case, so that for the user it would look like the original .NET exception bubbled through the whole stack. Though that would hide the wasm backtrace.)

However, I'm not sure how the error handling works: E.g. in the case when we return a host trap at the .NET callback, will the returned error at `wasmtime_func_call` always correspond to that set host trap, or could it also happen that another error is returned that is not related to the created trap? In that case, we might incorrectly store the .NET callback exception as InnerException of the `WasmtimeException` when the error has another cause than the trap.

What do you think?

Regarding the WASI exit status, I'm not sure how it can be retrieved now, as for calling `wasmtime_trap_exit_status` we would need to have a `wasm_trap_t`, not an `wasmtime_error_t`.

Additionally, I noticed that some tests in the CI seem to have failed due to an out of memory condition:
```
  Failed Wasmtime.Tests.Memory64AccessTests.ItThrowsForOutOfBoundsAccess [1 ms]
  Error Message:
   Wasmtime.WasmtimeException : Insufficient resources: VirtualAlloc failed: The paging file is too small for this operation to complete. (os error 1455)
```

Maybe the CI runner doesn't have enough RAM to always allocate a 4 GiB Memory? (Should we disable these Memory access tests for CI)?

Thanks!

**Comment by kpreisser** [comment]
> Regarding the WASI exit status, I'm not sure how it can be retrieved now, as for calling wasmtime_trap_exit_status we would need to have a wasm_trap_t*, not an wasmtime_error_t*.

This appears to be fixed with bytecodealliance/wasmtime#5215.
